runtime/include/MemRefUtils.hpp

```diff
@@ -0,0 +1,161 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstring>
+
+#include <Exception.hpp>
+
+extern "C" {
+void *_mlir_memref_to_llvm_alloc(size_t size);
+void *_mlir_memref_to_llvm_aligned_alloc(size_t alignment, size_t size);
+void _mlir_memref_to_llvm_free(void *ptr);
+}
+
+template <typename T, size_t R> struct MemRefT {
+    T *data_allocated;
+    T *data_aligned;
+    size_t offset;
+    size_t sizes[R];
+    size_t strides[R];
+};
+
+template <typename T, size_t R> class MemRefView {
+  private:
+    const MemRefT<T, R> *buffer;
+    size_t tsize; // total size
+
+  public:
+    class MemRefIter {
+      private:
+        const MemRefT<T, R> *buffer;
+        size_t loc; // physical index
+
+      public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T;
+        using difference_type = std::ptrdiff_t;
+        using pointer = T *;
+        using reference = T &;
+
+        MemRefIter(const MemRefT<T, R> *_buffer, size_t _idx = 0) : buffer(_buffer)
+        {
+            static_assert(R == 1, "[Class: MemRefIter] Error in Catalyst Runtime: Non-Iterable "
+                                  "MemRefs; 1-dim MemRefs are iterable.");
+            loc = buffer->offset + buffer->strides[0] * _idx;
+        }
+        pointer operator->() const { return &buffer->data_allocated[loc]; }
```

Comment:
I think this might be wrong. (It is usually correct for `data_allocated` == `data_aligned`)

I think this should be:

```suggestion
        pointer operator->() const { return &buffer->data_aligned[loc]; }
```

Reply:
:100:

Reply:
Just a note that the same extends to the next line.

Reply:
Yep, fixed it there already. I just wanted to have your suggestion there before my commits :)